File: 368-largest-divisible-subset/368-largest-divisible-subset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        if(nums.size()==1 || nums.size()==0)
        {
            return nums;
        }
        int n = nums.size();
        vector<int>dp(n,1);
        int maxlen = 1;
        for(int i=1;i<n;i++)
        {
            for(int j=0;j<i;j++)
            {
                if(nums[i]%nums[j]==0 && (1+dp[j]>dp[i]))
                {
                    dp[i] = 1 + dp[j];
                    maxlen = max(maxlen,dp[i]);
                }
            }
        }
        vector<int>ans;
        int prev;
        for(int i=n-1;i>=0;i--)
        {
            if(dp[i]==maxlen && (prev%nums[i])==0 || prev ==-1)
            {
                ans.push_back(nums[i]);
                maxlen--;
                prev = nums[i];
            }
        }
        return ans;
    }
};

File: 368-LargestDivisibleSubset/368-LargestDivisibleSubset.java
// Last updated: 11/2/2025, 10:32:07 PM
class Solution {

    public List<Integer> largestDivisibleSubset(int[] nums) {
        List<Integer> result = new ArrayList<>();
        Integer dp[] = new Integer[nums.length];

        Arrays.sort(nums);
        findSubSets(new ArrayList<>() ,result, -1, nums, 0, dp);
        return result;
    }

    public void findSubSets(List<Integer> curr, List<Integer> result, int prev, int[] nums, int i, Integer dp[]){
        if(i>=nums.length){
            if(curr.size() > result.size()){   
                result.clear();
                result.addAll(curr);
            }
            return;
        }

        if(dp[i]==null){
            dp[i] = -1;
        }
        
        
        
        if((prev ==-1 ||  nums[i]%prev==0) && curr.size()>dp[i]){
            dp[i] = curr.size();
            curr.add(nums[i]);
            findSubSets(curr, result, nums[i], nums, i+1, dp);
            curr.remove(curr.size()-1);
        }
        findSubSets(curr, result, prev, nums, i+1, dp);
    }
}